src/borg/compress.pyx

```diff
@@ -142,7 +147,38 @@ class LZ4(CompressorBase):
             osize = LZ4_compress_default(source, dest, isize, osize)
         if not osize:
             raise Exception('lz4 compress failed')
-        return super().compress(dest[:osize])
+        # only compress if the result actually is smaller
+        if osize < isize:
+            return self, dest[:osize]
+        else:
+            return NONE_COMPRESSOR, None
+
+    def decide(self, data):
+        return self._decide(data)[0]
+
+    def decide_compress(self, data):
+        """
+        Decides what to do with *data* and handle accordingly. Returns (compressor, compressed_data).
+
+        *compressed_data* is the result of *data* being processed by *compressor*.
+        """
+        compressor, compressed_data = self._decide(data)
+
+        if compressor is self:
+            return self, super().compress(compressed_data)
+
+        if compressed_data is None:
+            compressed_data = compressor.compress(data)
+
+        return compressor, compressed_data
+
+    def compress(self, data):
+        compressor, lz4_data = self._decide(data)
+
+        if compressor is self:
+            return super().compress(lz4_data)
+        else:
+            return compressor.compress(data)
```

Comment:
considering that this is basically `decide_compress(...)[0]`, could we move both to `BaseCompressor` class to avoid code duplication?

Reply:
Need to rethink this. Sounds good at first, but then either one also relies on the internal "decide and maybe compress" `_decide()` method, that would have to be present in `BaseCompressor` as well then. Also the current `super().compress()` idiom for "add compressor type header" would go away.
The most simple compressor CNONE uses these and would then have to optimize things by replacing those methods with simpler ones.
So would any potential new compressor that had a way to implement `decide()` without ever having to actually compress the data. The latter may be far fetched, but what I'm saying is, is that it doesn't feel right to give up the truely general nature of the base class and cause new code duplication to reduce another.

Maybe have one more layer of classes, ie. one like the current `BaseCompressor` that is general enough for a non-compressor like CNONE to be based on it without overhead, and another (like `SimpleCompressor` or some better name?) that adds the `decide()` based compression methods that are duplicate in the "normal" compressors (LZ4, LZMA, ZSD, any new algorithm)?

Reply:
hmm, sounds more complicated than estimated...

maybe try your suggestion in the last paragraph, but it's also fine if it doesn't work out to be worthwhile.

Reply:
Yeah, I will try it in a sub-branch of this one, then we can just ignore it at any point it turns out not to be worth the effort.

Meanwhile, I will push the other changes, plus another comment that needed updating due to the change, here to have them out the door.

Reply:
Splitting out the deuplicate code into a separate class did turn out quite straight forward, so its fine to go straight into this PR's branch. :slightly_smiling_face: